pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,337 @@
+	var matchedRule *CompletedRule
+	member := &crdv1beta.GroupMember{
+		Pod: &crdv1beta.PodReference{
+			Name:      podname,
+			Namespace: ns,
+		},
+	}
+	for ruleID := range item.ruleIDs {
+		// Iterate all rules relate to this multicast group address
+		r, _, _ := c.ruleCache.GetCompletedRule(ruleID)
+		// Find the rule with highest priority
+		if direction == r.Direction && r.TargetMembers.Has(member) == true {
+			if (matchedRule == nil) || *(item.ruleIDs[ruleID]) > *(item.ruleIDs[matchedRule.ID]) {
+				matchedRule = r
+			}
+		}
+	}
+	if matchedRule != nil {
+		ruleTypePtr = new(crdv1beta.NetworkPolicyType)
+		action, uuid, *ruleTypePtr, ruleName = *matchedRule.Action, matchedRule.PolicyUID, matchedRule.SourceRef.Type, matchedRule.Name
+	} else {
+		action, uuid, ruleName = v1alpha1.RuleActionAllow, "", ""
+	}
+	klog.V(2).InfoS("Call validation:", "action", action, "uuid", uuid, "ruleName", ruleName, "ruleType", ruleTypePtr)
+	return &action, uuid, ruleTypePtr, ruleName, nil
+}
+
+func (c *multicastController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *multicastController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect igmpRule to come off the workqueue.
+	if item, ok := obj.(*mcastIGMPRule); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected mcastIGMPRule in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncIGMPRule(item); err == nil {
+		// If no error occurs we Forget this item, it does not get queued again until
+		// another change happens.
+		c.queue.Forget(item)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(item)
+		klog.ErrorS(err, "error syncing multicast rule", "item", item)
+	}
+	return true
+}
+
+func (c *multicastController) syncIGMPRule(igmpRule *mcastIGMPRule) error {
+	c.ruleIDGroupMapMutex.Lock()
+	defer c.ruleIDGroupMapMutex.Unlock()
+	var item *mcastItem
+	switch igmpRule.event {
+	case ruleAdd:
+		klog.V(2).InfoS("Add rule", "igmpRule", igmpRule)
+		memberList := sets.String{}
+		for _, groupAddress := range igmpRule.groupAddress {
+			item = &mcastItem{
+				groupAddress: ip.IPv4StrToIPNet(groupAddress),
+				ruleIDs:      make(map[string]*uint16),
+			}
+			memberList.Insert(groupAddress)
+			obj, exists, _ := c.mcastItemRuleCache.GetByKey(groupAddress)
+			if exists {
+				// groupAddress exists
+				itemOrig := obj.(*mcastItem)
+				item.groupAddress = itemOrig.groupAddress
+				for ruleID, priority := range itemOrig.ruleIDs {
+					item.ruleIDs[ruleID] = priority
+				}
+			}
+			item.ruleIDs[igmpRule.ruleID] = igmpRule.priority
+
+			c.mcastItemRuleCache.Add(item)
+		}
+		c.ruleIDGroupAddressMap[igmpRule.ruleID] = memberList
```

Comment:
It seems `ruleIDGroupAddressMap` could just be an index of `mcastItemRuleCache`, and you actually created such an index but never used it.

Reply:
as discussed ,remove multicast controller, and use ruleCache in NetworkpolicyController